Probe for C11 thread local storage, and when available use it for "implicit context"

nwc10 opened the pull request:
Probe for compiler support for C11 thread local storage, and where available use this for "implicit context" for XS extensions making API calls for a threaded Perl build. This requires fewer function calls at the C level than POSIX thread specific storage. We continue to use the the pthreads approach if the C11 approach is not available.

Most operating systems ship a threaded Perl, and most XS extensions don't specify `PERL_NO_GET_CONTEXT`, hence they make a **lot** of calls to `pthread_getspecific` to retrieve the interpreter context to pass into each Perl API call.

Using thread local storage instead doesn't eliminate **all** function calls for a threaded shared object - there is still the need to make one call to get the TLS block. However, it drastically reduces the number of calls. For `APItest.xs` we go from

```
$ grep -c 'bl.*pthread_getspecific' ext/XS-APItest/APItest.s
8339
```

to

```
$ grep -c 'bl.*__tls_get_addr' ext/XS-APItest/APItest.s
3280
```

Curiously macOS doesn't support this even with current clang (due to the MACH-O executable format) , whilst most vendor Unix compilers from a decade ago support `__thread` even in C89 mode.

|OS, compiler|d_thread_local|perl_thread_local|
|------------|--------------|-----------------|
|HP-UX aCC   |define        |__thread         |
|HP-UX gcc   |define        |__thread         |
|AIX xlc     |undef         |                 |
|AIX gcc     |undef         |                 |
|Solaris suncc|define       |__thread         |
|Solaris gcc |define        |__thread         |

This was suncc 10.2 - I know that later versions default to C11 mode and support `_Thread_local`

nwc10 committed:
If we have thread local storage, use it instead of posix_getspecific().

Declare and use a variable PL_current_context to store a pointer to the
thread's interpreter struct. Use this to implement implicit context undef
ITHREADS, instead of a call to posix_getspecific().

For normal threaded code, this will eliminate function calls. For threaded
code in shared objects (the typical configuration for Linux distribution
builds) there still needs to be a function call to get the pointer to
thread local variables, but there will only be one per function, not one
per read of the variable.

nwc10 committed:
A Configure test for C11 thread local storage specificer, _Thread_local

We also provbe for gcc's earlier syntax, __thread.

nwc10 committed:
Update perldelta with the C11 thread local storage probe and usage.

nwc10 committed:
MACH-O can't support C11 thread local in shared objects, so disable it.

Apple's clang happily passes the probe and compiles _Thread_local no problem,
but building extensions fails with
    ld: illegal thread local variable reference to regular symbol _PL_current_context for architecture arm64

The Internet suggests that the MACH-O format fundamentally can't support
what is needed to implement C11 thread local storage in shared objects.
It's frustrating that the error message is "Less Than Awesome" at explaining
that this is the real problem here.

Hence disable the use of C11 thread local storage in the hints file, and
hence keep using pthreads.

xenu commented on win32/config.gc:
```diff
@@ -589,6 +589,7 @@ d_tcsetpgrp='undef'
 d_telldir='define'
 d_telldirproto='define'
 d_tgamma='define'
+d_thread_local='undef'
```
```suggestion
d_thread_local='define'
```

xenu commented on win32/config.gc:
```diff
@@ -961,6 +962,7 @@ perl5=''
 perl='perl'
 perl_patchlevel='~PERL_PATCHLEVEL~'
 perl_static_inline='static __inline__'
+perl_thread_local=''
```
```suggestion
perl_thread_local='__thread'
```

xenu commented on win32/config.vc:
```diff
@@ -589,6 +589,7 @@ d_tcsetpgrp='undef'
 d_telldir='define'
 d_telldirproto='define'
 d_tgamma='undef'
+d_thread_local='undef'
```
```suggestion
d_thread_local='define'
```

xenu commented on win32/config.vc:
```diff
@@ -960,6 +961,7 @@ perl5=''
 perl='perl'
 perl_patchlevel='~PERL_PATCHLEVEL~'
 perl_static_inline='static __inline'
+perl_thread_local=''
```
```suggestion
perl_thread_local='__declspec(thread)'
```

nwc10 replied:
I don't think that the code that I changed in `thread.h` is reachable for Win32. The Win32 code seems to be in `win32thread.h` (and already contains some conditional code to support `__declspec(thread)` that doesn't seem to be used.

Given that

1. I really don't know Win32
2. I don't have any access to Win32
3. I *think* that these code paths are all independent

it would seem better to me to leave these WIn32 config files as the default "disable", even if it's actually wrong, and then address the Win32 config in a separate PR.

Particularly given that the "fun" I already found (thank you Apple) on macOS.

So yes, agree this needs addressing, but can we do it next?

xenu replied:
Yeah, I guess I can prepare a PR after this one is merged (and actually test it this time :P)

nwc10 replied:
That would be cool

None of this stuff is obvious - I had a head start on you because i've been looking at it for longer, and I was forced to start by reading the entire files (rather than just looking at the changeset). And asking myself things like "What *is* `pthread_unchecked_getspecific_np` and does it matter?" (Answers seem to be "no good idea" and "probably not")

nwc10 replied:
So I think that this makes this "conversation" and the previous 3 "resolved"

nwc10 commented on configure.com:
```diff
@@ -6886,6 +6887,7 @@ $ WC "perllibs='" + perllibs + "'"
 $ WC "perlpath='" + "''vms_prefix':[000000]Perl''exe_ext'" + "'"
 $ WC "perl_static_inline='static inline'"
 $ WC "perl_symbol='" + perl_symbol + "'"  ! VMS specific
+$ WC "perl_thread_local=''" ! FIXME - as this is ia64 ABI, it may well be supported
```
@craigberry Given this whole `__thread` thing stared with ia64, does the VMS compiler support it? If so, like accurate win32 definitions, it might be easier to update this file in a later commit.

ilmari commented on thread.h:
```diff
@@ -379,19 +379,39 @@
-#ifndef PERL_GET_CONTEXT
-#  define PERL_GET_CONTEXT	PTHREAD_GETSPECIFIC(PL_thr_key)
-#endif
+#if defined(PERL_THREAD_LOCAL) && !defined(PERL_GET_CONTEXT) && !defined(PERL_GET_CONTEXT)
```
Duplicate `&& !defined(PERL_GET_CONTEXT)`?

nwc10 replied:
Oops yes. One should be PERL_SET_CONTEXT.

nwc10 replied:
Pushing an updated version with this fixed.

ilmari commented:
Have you done any benchmarking to see what the performance impact actually is in practice?

nwc10 commented:
> Have you done any benchmarking to see what the performance impact actually is in practice?

No. I was just trying to get the embuggerances unembuggered. Or however you say it. (Thank you Apple).

I wasn't really planning to. The generated assembler code is unambiguously simpler - I felt that this was enough to feel that it was worth the change. I don't have a good feel for whether it would be easily measurable, and if so what XS module(s) would be "typical" or "useful".

There's so much other PSC fun going on that I didn't really have the head space to figure this out.

nwc10 commented:
I think that this is good enough to go in. I don't think that it regresses on any platforms, and it benefits several.

Yes, we can (and should) improve Win32, and probably VMS, but those can be done independently. Also there might be minor improvements possible to the probe or implementation, but I don't think that there are any show stoppers.

@Tux if this goes in, it needs backporting to metaconfig. I hope I got everything roughly right. Also, it passes the "works on your machine" test - at least for the ones you're able to give access to.

nwc10 committed:
cygwin can't support C11 thread local in shared objects, so disable it.

This seems to be an unresolved bug in binutils, rather than any fundamental
object format problem.

nwc10 commented:
Removed the "Show Config" commit (which doesn't belong in **this** branch), and folded the *with USE_ITHREADS* commit into the *If we have thread local storage, use it* commit.

nwc10 commented:
mingw64 is made of fail, as ever, but otherwise we seem good.

jkeenan commented:
[snip]
> 
> ```
> $ grep -c 'bl.*pthread_getspecific' ext/XS-APItest/APItest.s
> 8339
> ```
> 
> to
> 
> ```
> $ grep -c 'bl.*__tls_get_addr' ext/XS-APItest/APItest.s
> 3280
> ```
> 

When I build perl on FreeBSD (generally threaded), I don't end up with any file named `ext/XS-APItest/APItest.s`.  
```
[perlmonger: XS-APItest] $ ls APItest\.* |cat
APItest.bs
APItest.c
APItest.o
APItest.pm
APItest.xs
```
Typo?

jkeenan commented:
If I apply your p.r., then configure once for a threaded build and once for unthreaded, then grep for the two new records in `config.sh`, I get:
```
config.sh.branch:617:d_thread_local='define'
config.sh.branch:993:perl_thread_local='_Thread_local'
config.sh.branch.unthreaded:615:d_thread_local='define'
config.sh.branch.unthreaded:991:perl_thread_local='_Thread_local'
```
Is that the expected result?

Thank you very much.
Jim Keenan

nwc10 commented:
> [snip]

> When I build perl on FreeBSD (generally threaded), I don't end up with any file named `ext/XS-APItest/APItest.s`.
 
> Typo?

No, you need to build it manually to inspect it. Sorry, I wasn't clear that this was a manual step:

```
make -C ext/XS-APItest/ APItest.s
```

nwc10 commented:
> If I apply your p.r., then configure once for a threaded build and once for unthreaded, then grep for the two new records in `config.sh`, I get:
> 
> ```
> config.sh.branch:617:d_thread_local='define'
> config.sh.branch:993:perl_thread_local='_Thread_local'
> config.sh.branch.unthreaded:615:d_thread_local='define'
> config.sh.branch.unthreaded:991:perl_thread_local='_Thread_local'
> ```
> 
> Is that the expected result?

Yes. (Again wasn't clear)

The probe was for whether the compiler supports it (it would take more work to make the probe conditional on ithreads).

We're only **using** it if we have ithreads:

```
#if defined(USE_ITHREADS) && defined(PERL_THREAD_LOCAL)
PERL_THREAD_LOCAL void *PL_current_context;
#endif
```

I guess it technically wouldn't even harm us to declare it even if we weren't using ithreads, but it would be a waste of space.

tonycoz commented:
Applied manually as 75595dd, a merge containing the individual commits
